src/Xamarin.Android.Build.Tasks/Utilities/NdkTools/NdkTools.cs

```diff
@@ -0,0 +1,391 @@
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Runtime.InteropServices;
+
+using Microsoft.Android.Build.Tasks;
+using Microsoft.Build.Utilities;
+using Xamarin.Android.Tools;
+
+namespace Xamarin.Android.Tasks
+{
+	public abstract class NdkTools
+	{
+		// Target triples used in various places (tools prefix, include directory naming etc)
+		static readonly Dictionary<AndroidTargetArch, string> archTriples = new Dictionary<AndroidTargetArch, string> {
+			{ AndroidTargetArch.Arm, "arm-linux-androideabi" },
+			{ AndroidTargetArch.Arm64, "aarch64-linux-android" },
+			{ AndroidTargetArch.X86, "i686-linux-android" },
+			{ AndroidTargetArch.X86_64, "x86_64-linux-android" },
+		};
+
+		// Architecture names as used by the `platforms/*` directories pre NDK r22
+		static readonly Dictionary<AndroidTargetArch, string> archPlatforms = new Dictionary<AndroidTargetArch, string> {
+			{ AndroidTargetArch.Arm, "arm" },
+			{ AndroidTargetArch.Arm64, "arm64" },
+			{ AndroidTargetArch.X86, "x86" },
+			{ AndroidTargetArch.X86_64, "x86_64" },
+		};
+
+		protected Dictionary<NdkToolKind, string> NdkToolNames = new Dictionary<NdkToolKind, string> {
+			{ NdkToolKind.Assembler, "as" },
+			{ NdkToolKind.Linker, "ld" },
+			{ NdkToolKind.Strip, "strip" },
+		};
+
+		public NdkVersion Version { get; }
+		public string NdkRootDirectory { get; }
+		public bool UsesClang { get; protected set; }
+		public bool NoBinutils { get; protected set; }
+
+		// We can't use MonoAndroidHelper.AndroidSdk.AndroidNdkHostPlatform here since it's
+		// not initialized while running tests and attempts to use cause a NREX
+		// We could call `MonoAndroidHelper.RefreshSupportedVersions` but since this NdkTools
+		// instance already knows the location of NDK, it would be just a waste of time.
+		protected string HostPlatform => GetNdkHostPlatform ();
+		protected bool IsWindows => OS.IsWindows;
+
+		protected TaskLoggingHelper? Log { get; }
+
+		protected NdkTools (string androidNdkPath, NdkVersion version, TaskLoggingHelper? log = null)
+		{
+			if (String.IsNullOrEmpty (androidNdkPath)) {
+				throw new ArgumentException ("must be a non-empty string", nameof (androidNdkPath));
+			}
+
+			NdkRootDirectory = androidNdkPath;
+			Version = version;
+		}
+
+		public static NdkTools? Create (string androidNdkPath, TaskLoggingHelper? log = null)
+		{
+			if (String.IsNullOrEmpty (androidNdkPath)) {
+				log?.LogCodedError ("XA5104", Properties.Resources.XA5104);
+				return null;
+			}
+
+			if (!Directory.Exists (androidNdkPath)) {
+				log?.LogCodedError ("XA5104", Properties.Resources.XA5104);
+				return null;
+			}
+
+			NdkVersion? version = ReadVersion (androidNdkPath, log);
+			if (version == null) {
+				return null;
+			}
+
+			if (version.Main.Major < 14) {
+				throw new InvalidOperationException ($"Unsupported NDK version {version}");
```

Comment:
This should also report an XA5104 error.  Sure, we "found" an NDK error, but as-is if someone tries to use NDK r13 (?! insanity ?!), the build will fail with the InvalidOperationException, which isn't as useful as an IOE + XA5104.